### PR TITLE
Add permission checks on global operations

### PR DESCRIPTION
**Pull request by Toflar**
This PR does 3 things:

- Rename the DC permissions and thus also the voter from `DCA` to `DC`. It's not voting on the "data container array" but on the "data container". So the A is incorrect.
- It also groups the permission attributes into
    * action permissions (essentially what our DC supports - `create`, `edit` etc.)
    * view permissions (essentially the operations ->  `global_operations`, `operations` and `buttons` which sometimes map to the action **but do not necessarily have to** which is why they must be separate permissions. Example: The `editheader` operation.
- It introduces permission checks in all those areas 🥳

### CODE REVIEW

File: core-bundle/src/Resources/config/services.yml [hunk 1]
```diff
@@ -746,8 +746,8 @@ services:
             - Contao\BackendUser
             - '@?logger'
 
-    contao.security.data_container.default_dca_voter:
-        class: Contao\CoreBundle\Security\Voter\DataContainer\DefaultDcaVoter
+    contao.security.data_container.default_dc_voter:
+        class: Contao\CoreBundle\Security\Voter\DataContainer\DefaultDcVoter
```
Comment (ausi) [inline core-bundle/src/Resources/config/services.yml]: Do we need the `DC` abbreviation here?
```suggestion
        class: Contao\CoreBundle\Security\Voter\DataContainer\DefaultDataContainerVoter
```

**Review by ausi — DISMISSED** [review]
Just one question regarding the naming of the default voter class. ☺️

**Comment by Toflar** [comment]
I don't mind either way :)

File: core-bundle/src/Resources/config/services.yml (hunk 1 above)
Comment (Toflar) [reply core-bundle/src/Resources/config/services.yml]: Done

**Comment by Toflar** [comment]
Will have to work a bit more on this after discussing with @aschempp. Our operations are not strictly related to our permissions. They are for e.g. `edit`, `delete` etc. but they might not match. E.g. `editheader` needs to check for `edit` of a different table. So we need to have different permissions on the operations :)

**Comment by Toflar** [comment]
This is ready for review now. Have a look at especially the `TODO` comments which I cannot really answer. I'd vote to merge this as is which means that in case you do not have access to an operation you would not have the icon anymore in the future. This would be a new UX which is totally fine for Contao 5 if we want this. If we still want to keep the UX of having disabled icons, then I'd still vote to merge this as is in order for us to be able to continue. Because what needs to follow would then be adjustments in the back end theme (e.g. automatically adding a CSS class that picks the "disabled" icon or whatever). The solution needed depends very much on the back end theme adjustments where I clearly cannot help out.

**Review by fritzmg — DISMISSED** [review]

File: core-bundle/src/Resources/contao/classes/DataContainer.php [hunk 2]
```diff
@@ -887,12 +887,24 @@ protected function generateButtons($arrRow, $strTable, $arrRootIds=array(), $bln
 
 		$return = '';
 
+		$security = System::getContainer()->get('security.helper');
+		$subject = new DataContainerSubject($strTable, $arrRow['id']);
```
Comment (ausi) [inline core-bundle/src/Resources/contao/classes/DataContainer.php]: I think we need a `rawurldecode()` here, see 5 lines below.
```suggestion
		$subject = new DataContainerSubject($strTable, rawurldecode($arrRow['id']));
```

File: core-bundle/src/Resources/contao/classes/DataContainer.php [hunk 3]
```diff
@@ -1134,6 +1146,9 @@ protected function generateHeaderButtons($arrRow, $strPtable)
 			return '';
 		}
 
+		$security = System::getContainer()->get('security.helper');
+		$subject = new DataContainerSubject($strPtable, $arrRow['id']);
```
Comment (ausi) [inline core-bundle/src/Resources/contao/classes/DataContainer.php]: Same here.
```suggestion
		$subject = new DataContainerSubject($strPtable, rawurldecode($arrRow['id']));
```

File: core-bundle/src/Resources/contao/drivers/DC_Folder.php [hunk 4]
```diff
@@ -1285,6 +1297,18 @@ public function move($blnIsAjax=false)
 			}
 		}
 
+		// Button permissions
+		$security = System::getContainer()->get('security.helper');
+		$subject = new DataContainerSubject($this->strTable); // Move view so we CAN provide an ID
```
Comment (ausi) [inline core-bundle/src/Resources/contao/drivers/DC_Folder.php]: The comment diverges from the code here. 
I think we should pass the id (which is the path) of the folder into which the user tries to upload to.

```suggestion
		$subject = new DataContainerSubject($this->strTable, $strFolder); // Move (upload) view so we CAN provide an ID
```

**Review by ausi — DISMISSED** [review]
Some minor findings. Apart from that ❤️

**Review by Toflar — COMMENTED** [review]
PR updated :)

**Review by ausi — DISMISSED** [review]

**Review by fritzmg — DISMISSED** [review]
> If we still want to keep the UX of having disabled icons, then I'd still vote to merge this as is in order for us to be able to continue. Because what needs to follow would then be adjustments in the back end theme (e.g. automatically adding a CSS class that picks the "disabled" icon or whatever). The solution needed depends very much on the back end theme adjustments where I clearly cannot help out.

In my opinion we should keep disabled icons (there is also a separate issue with a discussion on that already) - but yes, let's merge first and then adjust later 👍

**Comment by Toflar** [comment]
Yes, I agree. But what to do if the permission is not granted is a whole different discussion than actually implementing the checks :) That's why I think we should continue here so we can finally start transforming onload_callbacks into proper voters. What to do with the result of the voter can then be addressed separately.

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @Toflar.

**Pull request merged**